### PR TITLE
CPU model, topology and NUMA node tuning  GSoC

### PR DESCRIPTION
**Pull request by gqlo**
### What does this PR do?
In response to feature request #57880 

This PR add the feature to support CPU model and topology. configurable parameters are:
- `mode`
- `match`
- `check`
- `model`
- `vendor`
- `topology`
- `cache` 
- `feature` 

`numa` which is used to specify the attributes and sub-elements of numa `guest`. configurable parameters are:
- `cpus`
- `memory`
- `memAccess`
- `distance`
- `sibling` id and values

To configure numa `guest`,  numa `host ` must be configured beforehand, `numatune` can be used to specify `memory` which tells how to allocate memory for the domain process on a NUMA host and `memnode` which specify memory allocation policies per each guest NUMA node.

See [Libvirt domain XML](https://libvirt.org/formatdomain.html#elementsCPU) for more details.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/modules/virt.py [hunk 1]
```diff
@@ -740,35 +740,48 @@ def _gen_xml(
...
+                    for numa_id in context["cpu"]["numa"]:
+                        if context["cpu"]["numa"].get(numa_id):
+                            for k, v in context["cpu"]["numa"][numa_id].items():
+                                if k == "memory":
+                                    v = _handle_unit(v)
+                                if k != "distances":
+                                    context["cpu"]["numa"][numa_id].update(
+                                        {k: "{}='{}'".format(k, v)}
+                                    )
+        print("----------------context after ---------------------", context["cpu"])
```
Comment (cbosdo) [inline salt/modules/virt.py]: another debug print to be removed

File: salt/modules/virt.py [hunk 2]
```diff
@@ -1726,7 +1739,66 @@ def init(
...
+                   enabled: 'yes'
+                   hotpluggable: 'no'
+                   order: 1
+                 1:
+                   enabled: 'no'
+                   hotpluggable: 'yes'
+               attrib:
+                 match: minimum
+                 mode: custom
+                 check: full
```
Comment (cbosdo) [inline salt/modules/virt.py:1758]: I would move `match`, `mode` and `check` one level up to remove `attrib`

File: salt/modules/virt.py [hunk 3]
```diff
@@ -2164,6 +2296,7 @@ def seeder(path):
             numatune,
             **kwargs
         )
+        print("---------------------vm_xml-------------------------", vm_xml)
```
Comment (cbosdo) [inline salt/modules/virt.py:2293]: debug print to be removed or converted to `log.debug()`

**Review by cbosdo — CHANGES_REQUESTED** [review]
First round of comments. Code looks nice, may be simplified / factorized in some places and more important group commits to avoid the same code to change in every commit: makes the whole thing hard to read ;)

File: salt/modules/virt.py [hunk 4]
```diff
@@ -735,10 +736,58 @@ def _gen_xml(
...
+                    for vcpu_attr in context["cpu"]["vcpus"][vcpu_id]:
+                        context["cpu"]["vcpus"][vcpu_id].update(
+                            {
+                                vcpu_attr: "{}='{}'".format(
+                                    vcpu_attr,
+                                    context["cpu"]["vcpus"][vcpu_id][vcpu_attr],
+                                )
+                            }
+                        )
+            elif key in ["attrib", "model", "topology", "cache", "feature"]:
```
[the remark this answers is not in the document]
Comment (gqlo) [reply salt/modules/virt.py]: It's possible to get rid of `attrib`  But it might make the jinja code look even more ugly and less readable. This is how I handle `cpu attribute` in jinja templates  `<cpu {% for attr in cpu.attrib %} {{ cpu.attrib[attr] + ''}} {%- endfor %}>` There are 3 possible cpu attributes and 2 of them are optional, I can't think of a neater way of handle it. Do you have better suggestions?

[the remark this answers is not in the document]
Comment (gqlo) [reply salt/modules/virt.py]: Actually, i just had an idea of get rid of attrib without changing the jinja template :)

File: salt/modules/virt.py (hunk 2 above)
Comment (gqlo) [reply salt/modules/virt.py:1756]: got an idea already :)

**Comment by bryceml** [comment]
We should make sure the added options make it into the documentation.

**Comment by gqlo** [comment]
> We should make sure the added options make it into the documentation with example configuration.
@bryceml Hi Thanks for the comments. yeah i need to work on the documentation a bit more.

**Review by Akm0d — DISMISSED** [review]

**Review by cbosdo — CHANGES_REQUESTED** [review]
Refactoring in progress to make the code more generic. Do not merge as is

File: salt/modules/virt.py [hunk 5]
```diff
@@ -2757,10 +3179,8 @@ def _append_placeholder(value_dict, key):
                     for placeholder in placeholders
                 }
                 node_xpath = string.Template(xpath).substitute(ctx)
-
```
Comment (cbosdo) [inline salt/modules/virt.py:2430]: please avoid removing blank lines

File: tests/unit/modules/test_virt.py [hunk 6]
```diff
@@ -2715,6 +2819,176 @@ def test_update_existing_boot_params(self):
...
+
+        self.assertEqual(
+            setxml.find("./numatune/memnode/[@cellid='4']").attrib.get("mode"), "strict"
+        )
+
+        self.assertEqual(
+            setxml.find("./numatune/memnode/[@cellid='4']").attrib.get("nodeset"), "6"
+        )
+
+        self.assertEqual(setxml.find("./numatune/memnode/[@cellid='2']"), None)
```
Comment (cbosdo) [inline tests/unit/modules/test_virt.py:3807]: Why checking for a `memnode` that was never in the original nor updated description? better ensure that `cellid='1'` has been removed.

File: tests/unit/modules/test_virt.py [hunk 7]
```diff
@@ -2715,6 +2819,176 @@ def test_update_existing_boot_params(self):
...
+        self.assertEqual(
+            setxml.find("./numatune/memnode/[@cellid='3']").attrib.get("mode"),
+            "preferred",
+        )
+
+        self.assertEqual(
+            setxml.find("./numatune/memnode/[@cellid='3']").attrib.get("nodeset"), "7"
+        )
+
+        self.assertEqual(setxml.find("./numatune/memnode/[@cellid='2']"), None)
```
Comment (cbosdo) [inline tests/unit/modules/test_virt.py:3847]: This assert should be deleted too

File: tests/unit/modules/test_virt.py [hunk 8]
```diff
@@ -2715,6 +2819,176 @@ def test_update_existing_boot_params(self):
...
+
+        numatune_mnodes_none = {
+            "memory": {"mode": "preferred", "nodeset": "0-5"},
+            "memnodes": None,
+        }
```
Comment (cbosdo) [inline tests/unit/modules/test_virt.py:3852]: Instead of this test and the previous one, you should merge them into one that removes all numatune settings. It surely covers the same cases and would make it shorter

File: tests/unit/modules/test_virt.py [hunk 9]
```diff
@@ -2715,6 +2819,176 @@ def test_update_existing_boot_params(self):
...
+        )
+
+        self.assertEqual(
+            setxml.find("./numatune/memnode/[@cellid='4']").attrib.get("mode"), "strict"
+        )
+
+        self.assertEqual(
+            setxml.find("./numatune/memnode/[@cellid='4']").attrib.get("nodeset"), "6"
+        )
+
```
Comment (cbosdo) [inline tests/unit/modules/test_virt.py:3918]: I would remove that test since it's not that useful, and replace it by a test with exactly the parameters of the existing definition to ensure we're not updating the VM in such cases.

File: salt/modules/virt.py [hunk 10]
```diff
@@ -883,6 +882,76 @@ def _gen_xml(
...
+                                            if k == "discard":
+                                                context["cpu"]["numa"][cell_id][
+                                                    k
+                                                ] = "{}='{}'".format(
+                                                    k, "yes" if v else "no"
+                                                )
+                                            else:
+                                                context["cpu"]["numa"][cell_id][
+                                                    k
+                                                ] = "{}='{}'".format(k, v)
```
Comment (cbosdo) [inline salt/modules/virt.py:953]: I thought you implemented something simpler than that here... has it slipped in another PR?

File: salt/modules/virt.py [hunk 11]
```diff
@@ -1955,6 +2081,70 @@ def init(
...
             }
 
+    .. _init-cpu-def:
+
+    .. rubric:: cpu parameters definition
+
+    The cpu parameters dictionary can contain the following properties:
+
+    cpuset
+        a comma-separated list of physical CPU numbers that domain process and virtual CPUs can be pinned to by default
```
Comment (cbosdo) [inline salt/modules/virt.py:2091]: May be you should provide examples here. It also can handle ranges and excluded CPUs like `1-4,^3,6`

File: salt/modules/virt.py [hunk 12]
```diff
@@ -1955,6 +2081,70 @@ def init(
...
+
+    current
+        the number of virtual cpus available at startup
+
+    placement
+        indicate the CPU placement mode for domain process. the value can be either ``static`` or ``auto``
```
Comment (cbosdo) [inline salt/modules/virt.py:2097]: Shouldn't we make it easier by transforming into `auto-placement` as a boolean value defaulting to `True`?

File: salt/modules/virt.py [hunk 13]
```diff
@@ -1955,6 +2081,70 @@ def init(
...
+        ``mode`` attribute supported three possible values: ``emulate``, ``passthrough``, ``disable``
+
+    numa
+        specify the guest numa topology. ``cell`` element specifies a NUMA cell or a NUMA node, ``cpus`` specifies the
+        CPU or range of CPUs that are part of the node, ``memory`` specifies the node memory in kibibytes. All cells
+        should have ``id`` attribute in case referring to some cell is necessary in the code. optional attribute
+        ``memAccess`` control whether the memory is to be mapped as ``shared`` or ``private``, ``discard`` attribute which
+        fine tunes the discard feature for given numa node, possible values are ``yes`` or ``no``.  ``distances`` element
+        define the distance between NUMA cells and ``sibling`` sub-element is used to specify the distance value between
+        sibling NUMA cells.
```
Comment (cbosdo) [inline salt/modules/virt.py:2146]: Seems that documentation fragment is outdated already ;)

File: salt/modules/virt.py [hunk 14]
```diff
@@ -2528,7 +2718,7 @@ def _diff_graphics_lists(old, new):
 
 def update(
     name,
-    cpu=0,
+    cpu=None,
```
Comment (cbosdo) [inline salt/modules/virt.py:2721]: Do we really have to change the default value?

File: salt/modules/virt.py [hunk 15]
```diff
@@ -2809,6 +3006,184 @@ def _get_with_unit(node):
...
+            "set": _set_vcpu,
+        },
+        {"path": "cpu:maximum", "xpath": "vcpu", "get": lambda n: int(n.text)},
+        {
+            "path": "cpu:placement",
+            "xpath": "vcpu",
+            "get": lambda n: n.get("placement"),
+            "set": lambda n, v: n.set("placement", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("placement"),
+        },
```
Comment (cbosdo) [inline salt/modules/virt.py:3022]: If converting to `auto-placement` then this should be changed too

File: salt/modules/virt.py [hunk 16]
```diff
@@ -2809,6 +3006,184 @@ def _get_with_unit(node):
...
+            "set": lambda n, v: n.set("name", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("name"),
+        },
+        {
+            "path": "cpu:feature:policy",
+            "xpath": "cpu/feature",
+            "get": lambda n: n.get("policy"),
+            "set": lambda n, v: n.set("policy", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("policy"),
+        },
```
Comment (cbosdo) [inline salt/modules/virt.py:3126]: There is an issue here: there can (are likely to) be several features. We need to handle things like:
```
cpu:
  features:
    vmx: require
    lahf_lm: disable
```

File: tests/unit/modules/test_virt.py [hunk 17]
```diff
@@ -1841,13 +1841,282 @@ def test_update(self):
...
+                "interface": {"attached": [], "detached": []},
+            },
+            virt.update("my_vm", cpu=cpu_cache),
+        )
+        setxml = ET.fromstring(define_mock.call_args[0][0])
+        self.assertEqual(setxml.find("cpu").find("cache").attrib.get("level"), "3")
+        self.assertEqual(setxml.find("cpu").find("cache").attrib.get("mode"), "emulate")
+
+        # test adding feature
+        cpu_feature = {"feature": {"policy": "require", "name": "pcid"}}
```
Comment (cbosdo) [inline tests/unit/modules/test_virt.py:1988]: Please test adding multiple features instead.

File: tests/unit/modules/test_virt.py [hunk 18]
```diff
@@ -1841,13 +1841,282 @@ def test_update(self):
...
+        boot_uefi = {
+            "loader": "/usr/share/OVMF/OVMF_CODE.fd",
+            "nvram": "/usr/share/OVMF/OVMF_VARS.ms.fd",
+        }
+
+        invalid_boot = {
+            "loader": "/usr/share/OVMF/OVMF_CODE.fd",
+            "initrd": "/root/f8-i386-initrd",
+        }
+
```
Comment (cbosdo) [inline tests/unit/modules/test_virt.py:2119]: Hum, this is surely a rebase issue ;)

**Review by cbosdo — CHANGES_REQUESTED** [review]
There are still a few important things to fix in this PR, see the inline comments

File: salt/modules/virt.py [hunk 19]
```diff
@@ -1937,6 +1940,21 @@ def init(
 
         By default, the value will ``"hd"``.
 
+    :param numatune:
+        The optional numatune element provides details of how to tune the performance of a NUMA host via controlling NUMA
+        policy for domain process. The optional ``memory`` element specifies how to allocate memory for the domain process
+        on a NUMA host. ``memnode`` elements can specify memory allocation policies per each guest NUMA node. The definition
+        used in the dictionary can be found at :ref:`init-cpu-def`.
```
[the remark this answers is not in the document]
Comment (gqlo) [reply salt/modules/virt.py:2073]: hmm, actually `nodeset` `mode` are coming from `init-cpu-def`. They are somehow connected.

File: tests/unit/modules/test_virt.py (hunk 6 above)
Comment (gqlo) [reply tests/unit/modules/test_virt.py:3807]: I will add the assertion to ensure cellid='1' is removed. I guess it's still a good practice to check `memnode` that was never updated. Just in case there is something wrong with the update logic. We could run into the case that trying to update `cellid=1` which might messed with `cellid=2`. But i agree this is trivial.

File: salt/states/virt.py [hunk 20]
```diff
@@ -286,6 +286,7 @@ def defined(
     os_type=None,
     arch=None,
     boot=None,
+    numatune=None,
```
[the remark this answers is not in the document]
Comment (gqlo) [reply salt/states/virt.py:289]: thanks for pointing that out. I will add them

File: salt/modules/virt.py (hunk 15 above)
Comment (gqlo) [reply salt/modules/virt.py:3022]: It seems like we can pass `placement:auto` to indicate auto placement

File: tests/unit/modules/test_virt.py [hunk 21]
```diff
@@ -1871,7 +1871,7 @@ def test_update(self):
         )
         setxml = ET.fromstring(define_mock.call_args[0][0])
         self.assertEqual("2", setxml.find("vcpu").text)
-        self.assertEqual("2048", setxml.find("memory").text)
+        self.assertEqual("2147483648", setxml.find("memory").text)
```
Comment (s0undt3ch) [inline tests/unit/modules/test_virt.py:1874]: Can you please submit this fix in a separate PR.
I'd like to merge that sooner than this PR.
Thanks!

**Comment by gqlo** [comment]
re-run all

**Comment by gqlo** [comment]
re-run all

File: salt/modules/virt.py [hunk 22]
```diff
@@ -2578,6 +2791,7 @@ def update(
     graphics=None,
     live=True,
     boot=None,
+    numatune=None,
```
Comment (Ch3LL) [inline salt/modules/virt.py:2794]: this kwarg should be after `boot_dev` to safely cover the case for someone who is calling these kwargs as positional parameters.

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/modules/virt.py (hunk 22 above)
Comment (gqlo) [reply salt/modules/virt.py:2794]: That's right ! Thanks Megan.

File: salt/modules/virt.py (hunk 15 above)
Comment (cbosdo) [reply salt/modules/virt.py:3089]: Sure, but the idea here was to use a boolean rather than a string to avoid the user to look for the possible values in the doc.

File: salt/modules/virt.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (cbosdo) [reply salt/modules/virt.py:2135]: I am still convinced that the `numatune` parameter description shouldn't be shared with the cpu one, just for the sake of clarity. Otherwise you need to have clear references to the `init-cpu-def` sub section you mention. This starts to grow big, let's make it easy for the users.

**Review by cbosdo — CHANGES_REQUESTED** [review]
Still some doc changes to be made. Also look at https://github.com/cbosdo/salt/commit/90b6e0c8339b29d7c41f6502f2e911029095bf41 for a suggestion of how to better handle the `_gen_xml` added code.

**Comment by gqlo** [comment]
Commits of this PR is merged by another one.
